### ISSUE TITLE
partial matching of using $ for list elements can cause bugs

### ISSUE THREAD

**Issue by k-doering-NOAA** [r4ss/r4ss#442]
Thanks to @shcaba for reporting a recent example of this. (fixed in 96dfa2e)

It is often easier/faster to use `$` instead of `[[` to reference elements of lists in r4ss, because R studio will help you complete the expression. For example, `dat$CPUE` as opposed to `dat[["CPUE"]]`. These are generally equivalent, except that `$` allows partial matching while `[[` does not.

This can cause issues when list components can be NULL (thus, not produced), but partially matched with other existing list components. for example, `dat$CPUE` will match `dat$CPUEinfo` if the CPUE does not exist (i.e., is NULL). However, dat[["CPUE"]] will correctly return NULL.

This pops up occasionally creating bugs, but not that often. Some possibilities on how to deal with this:

1. Status quo: push fixes when these bugs come up.
2. Write a script that can parse the SS source code and replace all instances of $ with the `[[` formatting? Is there times when `$` does need to be used and changing would mess it up?
3. assign list components with NA, rather than conditionally creating them. I don't think there would be an automated way to do this, however, and it seems like it could be more work?

Anyone have thoughts on this?

**Comment by iantaylor-NOAA** [comment]
Option 2 sounds good to me although it would probably take a lot of manual checking to make sure the changes all made sense.

Regarding Option 3, I think that missing objects were NA long ago and I shifted to using NULL instead because 
`if (is.null(some.data.frame))` 
seemed cleaner than 
`if (!is.na(some.data.frame[[1]][1]))`
My sense is that it wouldn't be worth the effort to make the switch back to NA but if there would be other benefits beyond the partial matching problem, I could be convinced otherwise.

**Comment by roliveros-ramos** [comment]
There is another possibility for option 3, setting the elements to NULL:

```
dat = list(CPUEinfo=1:10)
dat$CPUE
```
[1]  1  2  3  4  5  6  7  8  9 10

```
dat = list(CPUEinfo=1:10, CPUE=NULL)
dat$CPUE
NULL
```
After a list has been created, names can be checked again a master names list and missing names added with NULL as value. That would eliminate the potential problems with partial matching. Or, start creating a list filled with names set to NULL.

**Comment by k-doering-NOAA** [comment]
Thanks @roliveros-ramos !  I've thought about that solution before, but I think sometimes when you set list values to NULL their names disappear from the list entirely? (Maybe you know more about the nuances of this?)

For example, if I run:
```r
dat <- list(CPUEinfo=1:10, CPUE=NULL)
dat[["CPUE"]] <- NULL
dat
```
I get back:
```
$CPUEinfo
 [1]  1  2  3  4  5  6  7  8  9 10
```
so the CPUE component (that is NULL) is no longer there...

**Comment by kellijohnson-NOAA** [comment]
I know that Ian and I have discussed this before but I cannot find the issue right now. I previously voted for [["xxx"]] because of the partial matching issue.

**Comment by roliveros-ramos** [comment]
@k-doering-NOAA, try this:

```
dat <- list(CPUEinfo=1:10, CPUE=NULL)
dat["CPUE"] <- list(NULL)
dat
```

to get:

```
$CPUEinfo
 [1]  1  2  3  4  5  6  7  8  9 10

$CPUE
NULL
```

Assigment to NULL is designed to remove an element of a list, which is a list itself. Assigning a list containing a NULL does the trick.

**Comment by k-doering-NOAA** [comment]
@roliveros-ramos thanks, I never knew that you could use `list(NULL)` to retain the name! Thank you for illustrating this distinction.

I am starting to think 2) would be the easiest path forward on fixing this specific issues, as I don't know if any examples in the r4ss package where `$` is intentionally used to allow partial matching...but I could be wrong. This would directly address the partial matching issues.

However,  I think 3) or the solution @roliveros-ramos put forward are worth considering in addition, and could be used in addition to 2). There may also be other benefits, like making all the names of elements that are missing more transparent to users.

**Comment by kellijohnson-NOAA** [comment]
Would there ever be an issue with a character symbol that can be used with the $ but not with [[? I have no idea if this is a thing or not, just trying to plan for the worst.

**Comment by k-doering-NOAA** [comment]
I was trying to think about this situation also, but couldn't think of an example?

According to the documentation in R (``` ?`$` ```), 
> x$name is equivalent to x[["name", exact = FALSE]]

so as long as I we are ok with everything being changed to x[["name", exact = **TRUE**]] (exact  = TRUE is the default), I don't think this would come up?

**Comment by k-doering-NOAA** [comment]
I started developing a script to read in a file and replace dollar signs in a new branch called issue_442. The first commit b5d16b9 includes some code to start and pseudocode needed to finish the function. Feel free to comment on the commit if you have any ideas about the approach or how it could be done better (I'm still learning about reg expressions and parsing strings).

**Comment by iantaylor-NOAA** [comment]
Thanks @k-doering-NOAA for working on this.

And thanks @roliveros-ramos for the tip about the `dat["CPUE"] <- list(NULL)` approach.

It's definitely appealing to have the transparency of the missing objects shown as NULL in the list created by `SS_output()`. The current code in that function creates a bunch of objects that are either NULL or not and then assigns them to a list (as in the following from https://github.com/r4ss/r4ss/blob/development/R/SS_output.R#L3421-L3426),

```
    # adding stuff to list which gets returned by function
    if (comp) {
      returndat$comp_data_exists <- TRUE
      returndat$lendbase <- lendbase
      returndat$sizedbase <- sizedbase
      returndat$agedbase <- agedbase 
```

From what I can tell, wrapping those objects in `list()` at the point where they are assigned to the bigger list is fine whether they are NULL or not.

**Comment by k-doering-NOAA** [comment]
Ok, I think I got the function figured out in most cases. I learned that you can use ```\\N``` (where N can be a digit 1-9) to literally backreference part of a regular expression in parentheses when using gsub. Here is some [more info on backreferences](https://www.regular-expressions.info/backref.html) . I figured this out from the awesome [rstudio regular expressions cheatsheet](https://rstudio.com/wp-content/uploads/2016/09/RegExCheatsheet.pdf).

The only case it won't work properly in that I'm aware of is if a reference has a nonsynactic name (in backticks) containing a dollar sign; for example, ```x$`bad$name` <- 55``` would incorrectly become `x[["bad[["name"]]"]] <- 55` . I think these names are pretty unlikely, though, so I have just added a warning (0df818c) when this happens rather than trying to stop it from happening (which seems like it would require a trickier solution...) 

I went ahead and ran this on SS_writedat_3.30 to see how it worked (4353a37) and I don't think there were any issues (although there wasn't anything terribly complicated either).

Feel free to comment on this issue if there are cases I have not accounted for!

Next steps:
- [ ] Figure out where this function should live. r4ss doesn't seem a good spot, as it really is more general (it could be used on any R code). Anyone have suggestions?
- [ ] Run this on the r4ss source code. The running part would be easy, but like @iantaylor-NOAA suggested, it may require some manual checking to feel certain it went as planned.

**Comment by k-doering-NOAA** [comment]
I ran the following code to replace $ in the r4ss source code:
```r
devtools::load_all(".")
r4ss_files <- list.files(path = "R", pattern = "*.[rR]", full.names = T)
for(r in r4ss_files) {
    r4ss:::rm_dollar_sign(file = r, allow_recursive = T, max_loops = 7)
}
```

I then committed all the changed files, except for the rm_dollar_sign.R file (I just didn't commit those changes b/c they are necessary for understanding the function).

commit 39223a6 includes these changes; I will look through and make sure the conversion went as it should, but anyone should feel free to point out issues if they see them!

**Comment by iantaylor-NOAA** [comment]
Thank you @k-doering-NOAA. 
I just glanced through some of the changes and everything looks good so far.
If the travis tests all past with these changes, then how about running styler::style_file() on all the files as well (as proposed in #399) to get these two large-scale style changes over with at the same time.

**Comment by k-doering-NOAA** [comment]
good suggestion @iantaylor-NOAA ! I did run styler on the same branch.

I did find one situation where the rm_dollar_sign function didn't work: using the dollar sign within strings, such as in message() (see fixes in dee3ae3e6f9c58894037be391680e84bf32d3ad3. Luckily, I think this is caught by trying to load or source the function, such as by running `devtools::load_all()`, so I don't think it is worth the effort to try to account for this scenario in rm_dollar_sign function.
